### ISSUE TITLE
closely-spaced bins on timeline make for very thin and sometimes invisible bars

### ISSUE THREAD

**Issue by joefutrelle** [WHOIGit/ifcbdb#86]
plotly is automagically computing the width of the bars. For irregularly-spaced bins, this creates the problem that the bars can become too narrow to see. Not an issue for other resolutions.

A workaround is to default to hours until the user is zoomed way in, but the solution is probably to explicitly set the bar width. It appears that plotly is capable of this, and tolerates the overlap it would cause.

**Comment by joefutrelle** [comment]
setting `bargap` to 0 mitigates this somewhat.

**Comment by joefutrelle** [comment]
it's actually possible in plotly js to set the width per-bar. we could potentially do this per bin using the bin timestamp for the left side of the bar and the bin timestamp + the run time for the right side.

https://codepen.io/etpinard/pen/RGmpgz

**Comment by joefutrelle** [comment]
hmm. except the above would not work if the sample time was well before the bin timestamp which it will be for discrete samples.

**Comment by mike-kaimika** [comment]
There's an example of the width parameter already being used in the code because it was required to fix a plotly bug if there is only one x axis value (otherwise Plotly won't render any bar):
`   // For bar graphs with only one data point, the width of the bar needs to be set explicitly or
    //   Plotly will not render anything visible to the user. The x range on a single entry plot is
    //   set to 24hours, so the bar is set to a width of 1 hour
    if (data["x"].length == 1) {
        series["width"] = [60*60*1000]
    }`

That code is setting the width to an hour because that's the resolution we're at I believe. But if we change the logic to be (bin timestamp) to (bin timestamp + run time), we might want to adjust this code as well

**Comment by joefutrelle** [comment]
Yeah, I'm still wondering what to do about sample times being different (earlier) than bin timestamps. In that case it doesn't make sense to make the bar extend from the sample time to the end of the run--although it might work visually to have the width be the run time even though we're displaying the bar at the sample time. Why don't we try making bin bars 20 minutes wide and see what happens with real data? It's not really a solution, but would give us some quick feedback that we could use to plan our eventual fix.
